toolkit/tools/pkg/imagecustomizerlib/cosicommon.go

```diff
@@ -0,0 +1,188 @@
+package imagecustomizerlib
+
+import (
+	"archive/tar"
+	"crypto/sha512"
+	_ "embed"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safeloopback"
+)
+
+type ImageBuildData struct {
+	Source    string
+	KnownInfo outputPartitionMetadata
+	Metadata  *Image
+}
+
+func convertToCosi(ic *ImageCustomizerParameters, imageUuid [UuidSize]byte, imageUuidStr string) error {
+	logger.Log.Infof("Extracting partition files")
+	outputDir := filepath.Join(ic.buildDir, "cosiimages")
+	err := os.MkdirAll(outputDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create folder %s:\n%w", outputDir, err)
+	}
+
+	imageLoopback, err := safeloopback.NewLoopback(ic.rawImageFile)
+	if err != nil {
+		return err
+	}
+	defer imageLoopback.Close()
+
+	partitionMetadataOutput, err := extractPartitions(imageLoopback.DevicePath(), outputDir, ic.outputImageBase, "raw-zst", imageUuid)
+	if err != nil {
+		return err
+	}
+
+	err = buildCosiFile(outputDir, ic.outputImageFile, partitionMetadataOutput, imageUuidStr)
+	if err != nil {
+		return fmt.Errorf("failed to build COSI: %w", err)
+	}
+
+	logger.Log.Infof("Successfully converted to COSI: %s", ic.outputImageFile)
+	return nil
+
+}
+
+func buildCosiFile(sourceDir string, outputFile string, expectedImages []outputPartitionMetadata, imageUuidStr string) error {
+	metadata := MetadataJson{
+		Version: "1.0",
+		OsArch:  runtime.GOARCH,
+		Id:      imageUuidStr,
+		Images:  make([]Image, len(expectedImages)),
+	}
+
+	if len(expectedImages) == 0 {
+		return errors.New("no images to build")
```

Comment:
Use `fmt.Errorf` instead of `errors.New`.

`errors.New` is used for global error objects, which are used when you want to support error type checking.